hadoop-tools/hadoop-azure/src/main/java/org/apache/hadoop/fs/azurebfs/AzureBlobFileSystemStore.java

```diff
@@ -890,32 +891,37 @@ public AbfsInputStream openFileForRead(Path path,
               encryptionContext.getBytes(StandardCharsets.UTF_8));
         }
       } else {
-        AbfsHttpOperation op = client.getPathStatus(relativePath, false,
-            tracingContext, null).getResult();
-        resourceType = op.getResponseHeader(
-            HttpHeaderConfigurations.X_MS_RESOURCE_TYPE);
-        contentLength = Long.parseLong(
-            op.getResponseHeader(HttpHeaderConfigurations.CONTENT_LENGTH));
-        eTag = op.getResponseHeader(HttpHeaderConfigurations.ETAG);
-        /*
-         * For file created with ENCRYPTION_CONTEXT, client shall receive
-         * encryptionContext from header field: X_MS_ENCRYPTION_CONTEXT.
-         */
-        if (client.getEncryptionType() == EncryptionType.ENCRYPTION_CONTEXT) {
-          final String fileEncryptionContext = op.getResponseHeader(
-              HttpHeaderConfigurations.X_MS_ENCRYPTION_CONTEXT);
-          if (fileEncryptionContext == null) {
-            LOG.debug("EncryptionContext missing in GetPathStatus response");
-            throw new PathIOException(path.toString(),
-                "EncryptionContext not present in GetPathStatus response headers");
+        if (client.getEncryptionType() == EncryptionType.ENCRYPTION_CONTEXT
```

Comment:
Good point. Actually, in case the encryptionType is ENCRYPTION_CONTEXT, read API needs to have following request headers:
- x-ms-encryption-key
- x-ms-encryption-key-sha256

To create these headers, encryptionContext is required which is given by the server on getPathStatus.

Hence, in case of ENCRYPTION_CONTEXT, we need to get the getPathStatus.

Reply:
Presently, say _abfsConfiguration.isInputStreamLazyOptimization_ is **disabled** then it will enter into the block even if its not ENCRYPTION_CONTEXT ? 

Appreciates if you could clarify - on what condition the below code should be skipped(can't be executed) ?

```
          final AbfsHttpOperation op = client.getPathStatus(relativePath, false,
              tracingContext, null).getResult();
          resourceType = op.getResponseHeader(
              HttpHeaderConfigurations.X_MS_RESOURCE_TYPE);
          contentLength = Long.parseLong(
              op.getResponseHeader(HttpHeaderConfigurations.CONTENT_LENGTH));
          eTag = op.getResponseHeader(HttpHeaderConfigurations.ETAG);
```

Reply:
I completely understand that logic has become bit messy and not easy to read. Have refactored the logic. Hope the new logic is more readable and understandable:

```
    if (fileStatus instanceof VersionedFileStatus) {
        VersionedFileStatus versionedFileStatus
            = (VersionedFileStatus) fileStatus;
        path = path.makeQualified(this.uri, path);
        Preconditions.checkArgument(fileStatus.getPath().equals(path),
            "Filestatus path [%s] does not match with given path [%s]",
            fileStatus.getPath(), path);
        resourceType = fileStatus.isFile() ? FILE : DIRECTORY;
        contentLength = fileStatus.getLen();
        eTag = versionedFileStatus.getVersion();
        encryptionContext = versionedFileStatus.getEncryptionContext();
      }

      if (client.getEncryptionType() == EncryptionType.ENCRYPTION_CONTEXT) {
        if (encryptionContext == null) {
          PathInformation pathInformation = getPathInformation(relativePath,
              tracingContext);
          resourceType = pathInformation.getResourceType();
          contentLength = Long.parseLong(pathInformation.getContentLength());
          eTag = pathInformation.getETag();
          encryptionContext = pathInformation.getEncryptionContext();

          if (encryptionContext == null) {
            LOG.debug("EncryptionContext missing in GetPathStatus response");
            throw new PathIOException(path.toString(),
                "EncryptionContext not present in GetPathStatus response headers");
          }
        }
        contextEncryptionAdapter = new ContextProviderEncryptionAdapter(
            client.getEncryptionContextProvider(), getRelativePath(path),
            encryptionContext.getBytes(StandardCharsets.UTF_8));
      } else if (fileStatus == null
          && !abfsConfiguration.isInputStreamLazyOptimizationEnabled()) {
        PathInformation pathInformation = getPathInformation(relativePath,
            tracingContext);
        resourceType = pathInformation.getResourceType();
        contentLength = Long.parseLong(pathInformation.getContentLength());
        eTag = pathInformation.getETag();
      }

      if (parseIsDirectory(resourceType)) {
        throw new AbfsRestOperationException(
            AzureServiceErrorCode.PATH_NOT_FOUND.getStatusCode(),
            AzureServiceErrorCode.PATH_NOT_FOUND.getErrorCode(),
            "openFileForRead must be used with files and not directories",
            null);
      }
```